exp/addons/internal/controllers/clusterresourceset_controller.go

```diff
@@ -280,17 +280,17 @@ func (r *ClusterResourceSetReconciler) ApplyClusterResourceSet(ctx context.Conte
 		return err
 	}
 
-	// Initialize the patch helper.
-	patchHelper, err := patch.NewHelper(clusterResourceSetBinding, r.Client)
-	if err != nil {
-		return err
-	}
+	patch := client.MergeFromWithOptions(clusterResourceSetBinding.DeepCopy(), client.MergeFromWithOptimisticLock{})
+
+	var errList []error
 
 	defer func() {
 		// Always attempt to Patch the ClusterResourceSetBinding object after each reconciliation.
-		if err := patchHelper.Patch(ctx, clusterResourceSetBinding); err != nil {
-			log.Error(err, "Failed to patch config")
+		if patchErr := r.Client.Patch(ctx, clusterResourceSetBinding, patch); patchErr != nil {
```

Comment:
I would prefer extending patchHelper.Patch to allow passing in the merge options as arguments.

Also: not using the patch helper would not patch status anymore.

Reply:
When doing it via opts and passing it only for CRS we won't have a change which breaks the signature, + we could use the same option if we (or folks importing the package) hit other issues which requires this.

Reply:
It isn't just a change to the signature, but a reasonably big refactor of the whole patch helper as the way that patch conditions, patch, and then patch status is applied doesn't pass the updated object to each call which is required for using optimistic locking as done in this PR. I feel that change would be unnecessarily large to fix this bug, compared to the size of the code fix in this PR, and could be done later if necessary for other controllers, which don't hit this bug due to their API semantics.

Of course, if you will only accept this being fixed by updating the patch helper then I will do that because this bug needs fixing, but I just wanted to share my feeling about it one last time 😅

Reply:
Similar to https://github.com/kubernetes-sigs/cluster-api/pull/10589#discussion_r1611675881:

> ... I would prefer using it so we can profit from error handling & potential improvements here as well (e.g. logging etc.)

Also it helps to reduce maintenance efforts.

We are already doing a Merge patch with `client.MergeFromWithOptimisticLock` always for conditions inside the patchhelper.
I could see doing it similarly, depending on if we pass in an `patch.Option` when calling `Patch(... patch.SpecWithOptimisticLocking{})`.

However, I agree, its not a small straight forward change, but it could be implemented to not impact the current way, used at other code paths.
So before digging into a refactor here it may be better to hear other opinions. cc @fabriziopandini @sbueringer

Reply:
If we go with the refactor of existing patch helper then I think the default (or even only option) should be optimistic locking and change the logic in the patch helper to just patch spec followed by status (which would include conditions). I think that should have a negligible effect on existing controllers but I can't be sure, and I don't have any context why the existing patch conditions first approach was done without applying optimistic locking generally.

The other option is switching to SSA for patch but that feels even more risky.

Reply:
I'm ok in using controller runtime patch + make the func fail in case of errors in the defer patch.
However probably we should get https://github.com/kubernetes-sigs/cluster-api/pull/10736 merged first

Reply:
@fabriziopandini just to confirm: you're ok with the approach in this pr?